Solutions/Marain.UserNotifications.Client/Marain/UserNotifications/Client/Management/Resources/CommunicationTemplates/SmsTemplateResource.cs

```diff
@@ -40,6 +40,23 @@ public SmsTemplateResource(in HalDocumentResource source)
         /// <inheritdoc/>
         public WebLink SelfLink => this.HalDocument.SelfLink;
 
+        /// <summary>
+        /// Gets the notification type of an SMS notification.
+        /// </summary>
+        public string NotificationType
+        {
+            get
+            {
+                if (this.HalDocument.TryGetProperty("notificationType", out JsonElement notificationType) &&
+                    notificationType.ValueKind == JsonValueKind.String)
+                {
+                    return notificationType.GetString();
+                }
+
+                throw new Exception("Schema violation - this value is required.");
```

Comment:
`Exception` is almost never the right type to throw. It should be something more specific. But I'm wondering if this is even the correct place to report this error.

Is there a reason not to check for this and throw it in the constructor? If it's illegal to have an `SmeTemplateResource` that doesn't include a notification type, we shouldn't really let people construct such a thing, and the ctor should throw an `ArgumentException` indicating that if someone tries.

If on the other hand we do in fact permit it, then it's not exactly a "schema violation". It's more just one possible state. And if we are saying "It's OK not to have a notification type, it's just that you mustn't ask what the notification type is when it's not there" then this should be an `InvalidOperationException`, and the error text should say something like "The NotificationType property must not be used on a template that does not specify a NotificationType".

Or this could return `string?`.

I don't know which of these options is correct, but what we have right now is wrong on two fronts: uninformative exception type, and a message that either blames the caller for a mistake that was actually made somewhere else (during construction), or it blames them for the wrong thing (attempting to use a property on an object that happens not to have that property available).

Reply:
Great spot, thanks for the writeup.

Reply:
To be honest, we copied this format from a pre-existing pattern in the rest of the file/project/other Marain project Clients. 

The validation I think can definitely be improved, but I think that would require a change of the rest of the file/other files in the project. The NotificationType is a "required value", so I guess the error does fit the message? Although it might not be to as how you expect it. Your statement could also be applied to other properties on the object.

I think the issue with changing it is more "What other questions will be asked if we try and change this, and how long will those delays be + the back and forth lag with that", we have a slight problem with that because we obviously have deadlines etc.

Reply:
Since this is a pre-existing issue with the client library, I've created #134 to deal with it.